README.md
BFPaperCheckbox
=============
[![CocoaPods](https://img.shields.io/cocoapods/v/BFPaperCheckbox.svg?style=flat)](https://github.com/bfeher/BFPaperCheckbox)

> iOS Checkboxes inspired by Google's Paper Material Design. 

![Animated Screenshot](https://raw.githubusercontent.com/bfeher/BFPaperCheckbox/master/BFPaperCheckboxDemoGif.gif "Animated Screenshot")


About
---------
### Now with Interface Builder customization support!

_BFPaperCheckbox_ is a subclass of UIControl that behaves much like the new paper checkboxes from Google's Material Design Labs.
All animation are asynchronous and are performed on sublayers.
BFPaperCheckboxes work right away with pleasing default behaviors, however they can be easily customized! The checkmark color and tap-circle color, both positive and negative (for checked and unchecked) are all readily customizable via public properties.
You can also set whether or not the tap-circle should appear from the location of the tap, or directly from the center of the control.

By default, BFPaperCheckboxes use "Smart Color" which will match the tap-circle's positive color to the color of the checkmark (`.checkmarkColor`).
You can set your own colors via: `.positiveColor` and `.negativeColor`. Note that setting these disables Smart Color functionality.

*Note: Try not to use super slow animation times. It breaks the visual effects (not the functionality though) plus, why would you?*


## Properties
`BOOL isChecked` <br />
> A BOOL representing the state of the checkbox. YES means checked, NO means unchecked.  

`CGFloat touchDownAnimationDuration` <br />
>(Go Steelers!) A CGFLoat representing the duration of the animations which take place on touch DOWN! Default is 0.25f seconds. Note that negative values will be converted to the default. NO FUNNY BUSINESS!  

`CGFloat touchUpAnimationDuration` <br />
>A CGFLoat representing the duration of the animations which take place on touch UP! Default is 2 * touchDownAnimationDuration seconds. Note that negative values will be converted to the default. NO FUNNY BUSINESS!  

`BOOL rippleFromTapLocation` <br />
>A flag to set to YES to have the tap-circle ripple from point of touch. If this is set to NO, the tap-circle will always ripple from the center of the button. Default is NO.  

`UIColor *checkmarkColor` <br />
>A UIColor to use for the checkmark color. Note that `self.tintColor` will be used for the square box color.  

`UIColor *positiveColor` <br />
>The UIColor to use for the circle which appears where you tap to check the box. Default value is smartly calculated from the `.checkmarkColor` property. Set to nil to access default. Alpha values less than 1 are recommended.  

`UIColor *negativeColor` <br />
>The UIColor to use for the circle which appears where you tap to uncheck the box. Default value is smartly calculated from the `.tintColor` property. Set to nil to access default. Alpha values less than 1 are recommended.  

`CGFloat cornerRadius` <br />
>The CGFloat value representing the corner radius of the control. Default value is (bfPaperCheckboxDefaultDiameter / 2). Note that negative values will be converted to 0. NO FUNNY BUSINESS!  

`CGFloat startDiameter` <br />
>A CGFLoat representing the diameter of the tap-circle as soon as it spawns, before it grows. Default is 1.f. Note that negative values and values less than 1 will be converted to the default. NO FUNNY BUSINESS!  

`CGFloat endDiameter` <br />
>The CGFloat value representing the Diameter of the tap-circle. By default it will take up the entire checkbox background circle. Note that zero and negative values will be converted to the default. NO FUNNY BUSINESS!  

`CGFloat burstAmount` <br />
>The CGFloat value representing how much we should increase the diameter of the tap-circle by when we burst it. Default is 0 because we can't see a burst with the default `.tapCircleDiameter` which takes up the entire frame. If you want to see a burst, make the `.tapCircleDiameter` something smaller than the diameter of the control itself. Note that negative values will be converted to the default. NO FUNNY BUSINESS!  

`id <BFPaperCheckboxDelegate> delegate` <br />
>A delegate to use our protocol with! The functions it conforms to are detailed below.  


## Delegate Functions
`(void)paperCheckboxChangedState:(BFPaperCheckbox *)checkbox`<br />
>An optional protocol method for detecting when the checkbox state changed. You can check its current state here with `checkbox.isChecked`.


## Constant
`CGFloat const bfPaperCheckboxDefaultDiameter`<br />
>A nice recommended value for size (49 points). (eg. `[[BFPaperCheckbox alloc] initWithFrame:CGRectMake(x, y, bfPaperCheckboxDefaultDiameter, bfPaperCheckboxDefaultDiameter)];`)


## Utility Functions (programmatically set state)
`(void)switchStatesAnimated:(BOOL)animated`<br />
>Use this function to manually/programmatically switch the state of this checkbox.
>@param `animated` A BOOL flag to choose whether or not to animate the change.

`(void)checkAnimated:(BOOL)animated`<br />
>Use this function to manually check the checkbox. Does nothing if already checked.
>@param `animated` A BOOL flag to choose whether or not to animate the change.  

`(void)uncheckAnimated:(BOOL)animated`<br />
>Use this function to manually uncheck the checkbox. Does nothing if already unchecked.
>@param `animated` A BOOL flag to choose whether or not to animate the change.



Usage
---------
Add the _BFPaperCheckbox_ header and implementation file to your project. (.h & .m)

### Creating a nice default BFPaperCheckbox
```objective-c
BFPaperCheckbox *paperCheckbox = [[BFPaperCheckbox alloc] initWithFrame:CGRectMake(x, y, bfPaperCheckboxDefaultDiameter, bfPaperCheckboxDefaultDiameter)];
```

### Customized Example
```objective-c
self.programmaticPaperCheckbox = [[BFPaperCheckbox alloc] initWithFrame:CGRectMake(0, 0, bfPaperCheckboxDefaultDiameter, bfPaperCheckboxDefaultDiameter)];
self.programmaticPaperCheckbox.delegate = self;
self.programmaticPaperCheckbox.tintColor = [UIColor colorWithRed:97.f/255.f green:97.f/255.f blue:97.f/255.f alpha:1];
self.programmaticPaperCheckbox.touchUpAnimationDuration = 0.5f;
self.programmaticPaperCheckbox.touchDownAnimationDuration = 0.5f;
self.programmaticPaperCheckbox.rippleFromTapLocation = YES;
self.programmaticPaperCheckbox.checkmarkColor = [UIColor blueColor];
self.programmaticPaperCheckbox.positiveColor = [[UIColor greenColor] colorWithAlphaComponent:0.5f];
self.programmaticPaperCheckbox.negativeColor = [[UIColor redColor] colorWithAlphaComponent:0.5f];
self.programmaticPaperCheckbox.startDiameter = 10;
self.programmaticPaperCheckbox.endDiameter = 35;
self.programmaticPaperCheckbox.burstAmount = 10;
[self.view addSubview:self.programmaticPaperCheckbox];
```

### Setting states manually
> Toggle with 'switchStates...'  
```objective-c
[self.paperCheckbox switchStatesAnimated:animate];
```  

> Manually check or uncheck:  
```objective-c
if (self.paperCheckbox.isChecked) {
  [self.paperCheckbox uncheckAnimated:animate];
} else {
  [self.paperCheckbox checkAnimated:animate];
}  
```  

CocoaPods
-------

CocoaPods are the best way to manage library dependencies in Objective-C projects.
Learn more at http://cocoapods.org

Add this to your podfile to add BFPaperCheckbox to your project.
```ruby
platform :ios, '7.0'
pod 'BFPaperCheckbox'
```


License
--------
_BFPaperCheckbox_ uses the MIT License:

> Please see included [LICENSE file](https://raw.githubusercontent.com/bfeher/BFPaperCheckbox/master/LICENSE.md).


CHANGELOG.md
BFPaperCheckbox
===============
[![CocoaPods](https://img.shields.io/cocoapods/v/BFPaperCheckbox.svg?style=flat)](https://github.com/bfeher/BFPaperCheckbox)

> Note that this changelog was started very late, at roughly the time between version 1.2.8 and 2.0.0 Non consecutive jumps in changelog mean that there were incremental builds that weren't released as a pod, typically while solving a problem.


2.1.0
---------
* (+) Added an IBInspectable 'cornerRadius' property.  


2.0.0
---------
* (+) IB Designable support!  
* (^) Replaced 'bfPaperCheckboxDefaultRadius' constant with a more sensible 'bfPaperCheckboxDefaultDiameter' constant.  
* (^) Lowered the default size of the control from 78 to 49, making it more compact.  
* (^) 'rippleFromTapLocation' now defaults to NO.  
* (^) 'tapCircleDiameter' has been renamed to 'endDiameter'  
* (^) 'tapCirclePositiveColor' has been renamed to 'positiveColor'  
* (^) 'tapCircleNegativeColor' has been renamed to 'negativeColor'  
* (+) Added new property 'touchDownAnimationDuration'  
* (+) Added new property 'touchUpAnimationDuration'  
* (+) Added new property 'startDiameter'  
* (+) Added new property 'burstAmount'  
* (-) Removed UIColor+BFPaperColors pod.

BFPaperCheckbox/BFPaperAppDelegate.h
//
//  BFPaperAppDelegate.h
//  BFPaperCheckbox
//
//  Created by Bence Feher on 7/22/14.
//  Copyright (c) 2014 Bence Feher. All rights reserved.
//
// The MIT License (MIT)
//
// Copyright (c) 2014 Bence Feher
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


#import <UIKit/UIKit.h>

@interface BFPaperAppDelegate : UIResponder <UIApplicationDelegate>

@property (strong, nonatomic) UIWindow *window;

@end


BFPaperCheckbox/BFPaperAppDelegate.m
//
//  BFPaperAppDelegate.m
//  BFPaperCheckbox
//
//  Created by Bence Feher on 7/22/14.
//  Copyright (c) 2014 Bence Feher. All rights reserved.
//
// The MIT License (MIT)
//
// Copyright (c) 2014 Bence Feher
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


#import "BFPaperAppDelegate.h"

@implementation BFPaperAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}
							
- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


BFPaperCheckbox/BFPaperViewController.h
//
//  BFPaperViewController.h
//  BFPaperCheckbox
//
//  Created by Bence Feher on 7/22/14.
//  Copyright (c) 2014 Bence Feher. All rights reserved.
//
// The MIT License (MIT)
//
// Copyright (c) 2014 Bence Feher
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


#import <UIKit/UIKit.h>

@interface BFPaperViewController : UIViewController

@end


BFPaperCheckbox/BFPaperViewController.m
//
//  BFPaperViewController.m
//  BFPaperCheckbox
//
//  Created by Bence Feher on 7/22/14.
//  Last updated by Bence Feher on 10/12/2016.
//  Copyright (c) 2016 Bence Feher. All rights reserved.
//
// The MIT License (MIT)
//
// Copyright (c) 2014 Bence Feher
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


#import "BFPaperViewController.h"
// Classes:
#import "BFPaperCheckbox.h"


@interface BFPaperViewController () <BFPaperCheckboxDelegate>
// Standard switch for comparison:
@property (weak, nonatomic) IBOutlet UISwitch *standardSwitch;
@property (weak, nonatomic) IBOutlet UILabel *standardSwitchLabel;
// BFPaperCheckboxes, both IB and programmatically created:
@property (weak, nonatomic) IBOutlet BFPaperCheckbox *storyboardPaperCheckbox;
@property (weak, nonatomic) IBOutlet UILabel *storyboardPaperCheckboxLabel;
@property BFPaperCheckbox *programmaticPaperCheckbox;
@property UILabel *programmaticPaperCheckboxLabel;
@end


@implementation BFPaperViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Set ourselves to tbe delegate of the standard switch:
    self.storyboardPaperCheckbox.delegate = self;   // Note that this guy is completely customized in the storyboard. Check out it's properties inspector and play around :)
    
    // Create and modify another BFPaperCheckbox programmatically:
    self.programmaticPaperCheckbox = [[BFPaperCheckbox alloc] initWithFrame:CGRectMake(0, 0, bfPaperCheckboxDefaultDiameter, bfPaperCheckboxDefaultDiameter)];
    self.programmaticPaperCheckbox.tag = 1002;
    self.programmaticPaperCheckbox.delegate = self;
    self.programmaticPaperCheckbox.tintColor = [UIColor colorWithRed:97.f/255.f green:97.f/255.f blue:97.f/255.f alpha:1];
    // Note, we will use default values for the rest, change them yourself to experiment and find a setup you like :)
    self.programmaticPaperCheckbox.touchUpAnimationDuration = -1;
    self.programmaticPaperCheckbox.touchDownAnimationDuration = -1;
    self.programmaticPaperCheckbox.rippleFromTapLocation = NO;
    self.programmaticPaperCheckbox.checkmarkColor = nil;
    self.programmaticPaperCheckbox.positiveColor = nil;
    self.programmaticPaperCheckbox.negativeColor = nil;
    self.programmaticPaperCheckbox.startDiameter = -1;
    self.programmaticPaperCheckbox.endDiameter = 0;
    self.programmaticPaperCheckbox.burstAmount = 0;
    
    [self.view addSubview:self.programmaticPaperCheckbox];
    
    // Setup the label for our programmaticPaperCheckbox:
    self.programmaticPaperCheckboxLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 286, 21)];
    self.programmaticPaperCheckboxLabel.minimumScaleFactor = 0.5f;
    self.programmaticPaperCheckboxLabel.text = @"Programmatic [OFF]";
    [self.view addSubview:self.programmaticPaperCheckboxLabel];
    
    
    // Set up a standard switch:
    self.standardSwitch.tintColor = [UIColor colorWithRed:97.f/255.f green:97.f/255.f blue:97.f/255.f alpha:1];
    self.standardSwitch.onTintColor = [UIColor colorWithRed:76.f/255.f green:175.f/255.f blue:80.f/255.f alpha:1];
    [self.standardSwitch addTarget:self action:@selector(standardSwitchChangedState:) forControlEvents:UIControlEventTouchUpInside];
    
    
    // Autolayout code; you can ignore this.
    self.programmaticPaperCheckbox.translatesAutoresizingMaskIntoConstraints = NO;
    self.programmaticPaperCheckboxLabel.translatesAutoresizingMaskIntoConstraints = NO;
    NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:self.programmaticPaperCheckbox
                                                                       attribute:NSLayoutAttributeWidth
                                                                       relatedBy:NSLayoutRelationEqual
                                                                          toItem:nil
                                                                       attribute:NSLayoutAttributeNotAnAttribute
                                                                      multiplier:1 constant:bfPaperCheckboxDefaultDiameter];
    NSLayoutConstraint *heightConstraint = [NSLayoutConstraint constraintWithItem:self.programmaticPaperCheckbox
                                                                        attribute:NSLayoutAttributeHeight
                                                                        relatedBy:NSLayoutRelationEqual
                                                                           toItem:nil
                                                                        attribute:NSLayoutAttributeNotAnAttribute
                                                                       multiplier:1 constant:bfPaperCheckboxDefaultDiameter];
    [self.programmaticPaperCheckbox addConstraints:@[widthConstraint, heightConstraint]];
    NSLayoutConstraint *centerXConstraint = [NSLayoutConstraint constraintWithItem:self.programmaticPaperCheckbox
                                                                         attribute:NSLayoutAttributeCenterX
                                                                         relatedBy:NSLayoutRelationEqual
                                                                            toItem:self.storyboardPaperCheckbox
                                                                         attribute:NSLayoutAttributeCenterX
                                                                        multiplier:1
                                                                          constant:0];
    NSLayoutConstraint *verticalSpacingConstraint = [NSLayoutConstraint constraintWithItem:self.programmaticPaperCheckbox
                                                                                 attribute:NSLayoutAttributeTop
                                                                                 relatedBy:NSLayoutRelationEqual
                                                                                    toItem:self.storyboardPaperCheckbox
                                                                                 attribute:NSLayoutAttributeBottom
                                                                                multiplier:1 constant:18];
    NSLayoutConstraint *centerYConstraint = [NSLayoutConstraint constraintWithItem:self.programmaticPaperCheckboxLabel
                                                                         attribute:NSLayoutAttributeCenterY
                                                                         relatedBy:NSLayoutRelationEqual
                                                                            toItem:self.programmaticPaperCheckbox
                                                                         attribute:NSLayoutAttributeCenterY
                                                                        multiplier:1 constant:0];
    NSLayoutConstraint *leadingLabelConstraint = [NSLayoutConstraint constraintWithItem:self.programmaticPaperCheckboxLabel
                                                                              attribute:NSLayoutAttributeLeading
                                                                              relatedBy:NSLayoutRelationEqual
                                                                                 toItem:self.programmaticPaperCheckbox
                                                                              attribute:NSLayoutAttributeTrailing
                                                                             multiplier:1 constant:8];
    NSLayoutConstraint *trailingLabelConstraint = [NSLayoutConstraint constraintWithItem:self.programmaticPaperCheckboxLabel
                                                                               attribute:NSLayoutAttributeTrailing
                                                                               relatedBy:NSLayoutRelationEqual
                                                                                  toItem:self.view
                                                                               attribute:NSLayoutAttributeTrailingMargin
                                                                              multiplier:1 constant:0];
    [self.view addConstraints:@[centerXConstraint, verticalSpacingConstraint, centerYConstraint, leadingLabelConstraint, trailingLabelConstraint]];
}


- (IBAction)tapped:(UIButton *)sender {
    BOOL animate;
    if (sender.tag == 2001) {       // animate button tag
        animate = YES;
    }
    else if (sender.tag == 2002) {  // static button tag
        animate = NO;
    }
    // Animate our standard switch:
    if (self.standardSwitch.isOn) {
        [self.standardSwitch setOn:NO animated:animate];
    } else {
        [self.standardSwitch setOn:YES animated:animate];
    }
    [self standardSwitchChangedState:self.standardSwitch];

    /*
     * Below are the two ways of programmatically setting the state of a checkbox.
     */
    // (1) Swap storyboardPaperCheckbox state with the 'switchStates...' method:
    [self.storyboardPaperCheckbox switchStatesAnimated:animate];
    
    // (2) Swap programmaticPaperCheckbox's state with the 'check...'/'uncheck...' methods:
    if (self.programmaticPaperCheckbox.isChecked) {
        [self.programmaticPaperCheckbox uncheckAnimated:animate];
    }
    else {
        [self.programmaticPaperCheckbox checkAnimated:animate];
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


#pragma mark UISwitch Delegate
- (IBAction)standardSwitchChangedState:(UISwitch *)sender {
    self.standardSwitchLabel.text = sender.isOn ? @"Standard switch [ON]" : @"Standard switch [OFF]";
}

#pragma mark - BFPaperCheckbox Delegate
- (void)paperCheckboxChangedState:(BFPaperCheckbox *)checkbox
{
    if (checkbox.tag == 1001) {      // First box
        self.storyboardPaperCheckboxLabel.text = self.storyboardPaperCheckbox.isChecked ? @"IB Custom [ON]" : @"IB Custom [OFF]";
    }
    else if (checkbox.tag == 1002) { // Second box
        self.programmaticPaperCheckboxLabel.text = self.programmaticPaperCheckbox.isChecked ? @"Programmatic [ON]" : @"Programmatic [OFF]";
    }
}

@end


Classes/BFPaperCheckbox.h
//
//  BFPaperCheckbox.h
//  BFPaperCheckbox
//
//  Created by Bence Feher on 7/22/14.
//  Last updated by Bence Feher on 10/12/2016.
//  Copyright (c) 2016 Bence Feher. All rights reserved.
//
// The MIT License (MIT)
//
// Copyright (c) 2014 Bence Feher
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


#import <UIKit/UIKit.h>
#import <QuartzCore/QuartzCore.h>

@class BFPaperCheckbox;
@protocol BFPaperCheckboxDelegate <NSObject>
@optional
/** An optional protocol method for detecting when the checkbox state changed. You can check its current state here with 'checkbox.isChecked'. */
- (void)paperCheckboxChangedState:(BFPaperCheckbox *)checkbox;
@end


IB_DESIGNABLE
#if __IPHONE_OS_VERSION_MAX_ALLOWED < 100000
// CAAnimationDelegate is not available before iOS 10 SDK
@interface BFPaperCheckbox : UIControl <UIGestureRecognizerDelegate>
#else
@interface BFPaperCheckbox : UIControl <UIGestureRecognizerDelegate,CAAnimationDelegate>
#endif

#pragma mark - Constants
/** A nice recommended value for size. (eg. [[BFPaperCheckbox alloc] initWithFrame:CGRectMake(x, y, bfPaperCheckboxDefaultDiameter, bfPaperCheckboxDefaultDiameter)]; */
extern CGFloat const bfPaperCheckboxDefaultDiameter;


#pragma mark - Properties
#pragma mark Animation
/** (Go Steelers!) A CGFLoat representing the duration of the animations which take place on touch DOWN! Default is 0.25f seconds.  Note that negative values will be converted to the default. NO FUNNY BUSINESS! */
@property IBInspectable CGFloat touchDownAnimationDuration;
/** A CGFLoat representing the duration of the animations which take place on touch UP! Default is 2 * touchDownAnimationDuration seconds. Note that negative values will be converted to the default. NO FUNNY BUSINESS! */
@property IBInspectable CGFloat touchUpAnimationDuration;

/** A flag to set to YES to have the tap-circle ripple from point of touch. If this is set to NO, the tap-circle will always ripple from the center of the button. Default is NO. */
@property IBInspectable BOOL rippleFromTapLocation;

#pragma mark Colors
/** A UIColor to use for the checkmark color. Note that 'self.tintColor' will be used for the square box color. */
@property IBInspectable UIColor *checkmarkColor;

/** The UIColor to use for the circle which appears where you tap to check the box. Default value is calculated from the '.checkmarkColor' property. Set to nil to access default. Alpha values less than 1 are recommended. */
@property IBInspectable UIColor *positiveColor;

/** The UIColor to use for the circle which appears where you tap to uncheck the box. Default value is calculated from the '.tintColor' property. Set to nil to access default. Alpha values less than 1 are recommended. */
@property IBInspectable UIColor *negativeColor;

#pragma mark Sizing
/** The CGFloat value representing the corner radius of the control. Default value is (bfPaperCheckboxDefaultDiameter / 2). Note that negative values will be converted to 0. NO FUNNY BUSINESS! */
@property (nonatomic) IBInspectable CGFloat cornerRadius;

/** A CGFLoat representing the diameter of the tap-circle as soon as it spawns, before it grows. Default is 1.f. Note that negative values and values less than 1 will be converted to the default. NO FUNNY BUSINESS! */
@property IBInspectable CGFloat startDiameter;

/** The CGFloat value representing the Diameter of the tap-circle. By default it will take up the entire checkbox background circle. Note that zero and negative values will be converted to the default. NO FUNNY BUSINESS! */
@property IBInspectable CGFloat endDiameter;

/** The CGFloat value representing how much we should increase the diameter of the tap-circle by when we burst it. Default is 0 because we can't see a burst with the default '.tapCircleDiameter' which takes up the entire frame. If you want to see a burst, make the '.tapCircleDiameter' something smaller than the diameter of the control itself. Note that negative values will be converted to the default. NO FUNNY BUSINESS! */
@property IBInspectable CGFloat burstAmount;

#pragma mark Status
/** A BOOL representing the state of the checkbox. YES means checked, NO means unchecked. **/
@property (nonatomic, readonly) BOOL isChecked;

#pragma mark Delegate
/** A delegate to use our protocol with! */
@property (weak) id <BFPaperCheckboxDelegate> delegate;


#pragma mark - Utility Functions
/**
 *  Use this function to manually/programmatically switch the state of this checkbox.
 *
 *  @param animated A BOOL flag to choose whether or not to animate the change.
 */
- (void)switchStatesAnimated:(BOOL)animated;

/**
 *  Use this function to manually check the checkbox. Does nothing if already checked.
 *
 *  @param animated A BOOL flag to choose whether or not to animate the change.
 */
- (void)checkAnimated:(BOOL)animated;

/**
 *  Use this function to manually uncheck the checkbox. Does nothing if already unchecked.
 *
 *  @param animated A BOOL flag to choose whether or not to animate the change.
 */
- (void)uncheckAnimated:(BOOL)animated;
@end
